// @octokit/webhooks-methods/dist-node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var crypto = require('crypto');
var buffer = require('buffer');

const VERSION = "1.0.0";

var Algorithm;

(function (Algorithm) {
  Algorithm["SHA1"] = "sha1";
  Algorithm["SHA256"] = "sha256";
})(Algorithm || (Algorithm = {}));

async function sign(options, payload) {
  const {
    secret,
    algorithm
  } = typeof options === "object" ? {
    secret: options.secret,
    algorithm: options.algorithm || Algorithm.SHA256
  } : {
    secret: options,
    algorithm: Algorithm.SHA256
  };

  if (!secret || !payload) {
    throw new TypeError("[@octokit/webhooks-methods] secret & payload required for sign()");
  }

  if (!Object.values(Algorithm).includes(algorithm)) {
    throw new TypeError(`[@octokit/webhooks] Algorithm ${algorithm} is not supported. Must be  'sha1' or 'sha256'`);
  }

  payload = typeof payload === "string" ? payload : toNormalizedJsonString(payload);
  return `${algorithm}=${crypto.createHmac(algorithm, secret).update(payload).digest("hex")}`;
}

function toNormalizedJsonString(payload) {
  return JSON.stringify(payload).replace(/[^\\]\\u[\da-f]{4}/g, s => {
    return s.substr(0, 3) + s.substr(3).toUpperCase();
  });
}

sign.VERSION = VERSION;

const getAlgorithm = signature => {
  return signature.startsWith("sha256=") ? "sha256" : "sha1";
};

async function verify(secret, eventPayload, signature) {
  if (!secret || !eventPayload || !signature) {
    throw new TypeError("[@octokit/webhooks-methods] secret, eventPayload & signature required");
  }

  const signatureBuffer = buffer.Buffer.from(signature);
  const algorithm = getAlgorithm(signature);
  const verificationBuffer = buffer.Buffer.from(await sign({
    secret,
    algorithm
  }, eventPayload));

  if (signatureBuffer.length !== verificationBuffer.length) {
    return false;
  } // constant time comparison to prevent timing attachs
  // https://stackoverflow.com/a/31096242/206879
  // https://en.wikipedia.org/wiki/Timing_attack


  return crypto.timingSafeEqual(signatureBuffer, verificationBuffer);
}
verify.VERSION = VERSION;

exports.sign = sign;
exports.verify = verify;
