// octokit/dist-node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@octokit/core');
var pluginPaginateRest = require('@octokit/plugin-paginate-rest');
var pluginRestEndpointMethods = require('@octokit/plugin-rest-endpoint-methods');
var pluginRetry = require('@octokit/plugin-retry');
var pluginThrottling = require('@octokit/plugin-throttling');
var app = require('@octokit/app');
var oauthApp = require('@octokit/oauth-app');

const VERSION = "1.0.3";

const Octokit = core.Octokit.plugin(pluginRestEndpointMethods.restEndpointMethods, pluginPaginateRest.paginateRest, pluginRetry.retry, pluginThrottling.throttling).defaults({
  userAgent: `octokit-rest.js/${VERSION}`,
  throttle: {
    onRateLimit,
    onAbuseLimit
  }
}); // istanbul ignore next no need to test internals of the throttle plugin

function onRateLimit(retryAfter, options, octokit) {
  octokit.log.warn(`Request quota exhausted for request ${options.method} ${options.url}`);

  if (options.request.retryCount === 0) {
    // only retries once
    octokit.log.info(`Retrying after ${retryAfter} seconds!`);
    return true;
  }
} // istanbul ignore next no need to test internals of the throttle plugin


function onAbuseLimit(retryAfter, options, octokit) {
  octokit.log.warn(`Abuse detected for request ${options.method} ${options.url}`);

  if (options.request.retryCount === 0) {
    // only retries once
    octokit.log.info(`Retrying after ${retryAfter} seconds!`);
    return true;
  }
}

const App = app.App.defaults({
  Octokit
});
const OAuthApp = oauthApp.OAuthApp.defaults({
  Octokit
});

exports.App = App;
exports.OAuthApp = OAuthApp;
exports.Octokit = Octokit;
