// octokit/dist-web/index.js
import { Octokit as Octokit$1 } from '@octokit/core';
import { paginateRest } from '@octokit/plugin-paginate-rest';
import { restEndpointMethods } from '@octokit/plugin-rest-endpoint-methods';
import { retry } from '@octokit/plugin-retry';
import { throttling } from '@octokit/plugin-throttling';
import { App as App$1 } from '@octokit/app';
import { OAuthApp as OAuthApp$1 } from '@octokit/oauth-app';

const VERSION = "1.0.3";

const Octokit = Octokit$1.plugin(restEndpointMethods, paginateRest, retry, throttling).defaults({
    userAgent: `octokit-rest.js/${VERSION}`,
    throttle: {
        onRateLimit,
        onAbuseLimit,
    },
});
// istanbul ignore next no need to test internals of the throttle plugin
function onRateLimit(retryAfter, options, octokit) {
    octokit.log.warn(`Request quota exhausted for request ${options.method} ${options.url}`);
    if (options.request.retryCount === 0) {
        // only retries once
        octokit.log.info(`Retrying after ${retryAfter} seconds!`);
        return true;
    }
}
// istanbul ignore next no need to test internals of the throttle plugin
function onAbuseLimit(retryAfter, options, octokit) {
    octokit.log.warn(`Abuse detected for request ${options.method} ${options.url}`);
    if (options.request.retryCount === 0) {
        // only retries once
        octokit.log.info(`Retrying after ${retryAfter} seconds!`);
        return true;
    }
}

const App = App$1.defaults({ Octokit });
const OAuthApp = OAuthApp$1.defaults({ Octokit });

export { App, OAuthApp, Octokit };
